streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -95,6 +97,7 @@
     private static final String JMX_PREFIX = "kafka.streams";
     public static final int DEFAULT_CLOSE_TIMEOUT = 0;
     private final StreamThread[] threads;
+    private final Map<Long, StreamThread.State> threadState;
```

Comment:
Why is `synchronized(stateListener)` wrong? There are two ways of doing it, this way synchronises on the common object these threads have. I agree one can also synchronize on the method `onChange`. I don't immediately see why one is better than the other.